ext/priority_sampling/priority_sampling.c

```diff
@@ -81,6 +106,28 @@ static void dd_decide_on_sampling(ddtrace_span_data *span) {
         }
         ZEND_HASH_FOREACH_END();
 
+        if (!explicit_rule) {
+            ddtrace_try_read_agent_rate();
+
+            if (DDTRACE_G(agent_rate_by_service)) {
+                zval *env = zend_hash_str_find(ddtrace_spandata_property_meta(span), ZEND_STRL("env"));
+                zval *sample_rate_zv = NULL;
+                if (Z_TYPE_P(service) == IS_STRING && env && Z_TYPE_P(env) == IS_STRING) {
+                    zend_string *sample_key = zend_strpprintf(0, "service:%.*s,env:%.*s",(int) Z_STRLEN_P(service), Z_STRVAL_P(service),
+                                                              (int) Z_STRLEN_P(env), Z_STRVAL_P(env));
+                    sample_rate_zv = zend_hash_find(DDTRACE_G(agent_rate_by_service), sample_key);
+                    zend_string_release(sample_key);
+                }
+                if (!sample_rate_zv) {
+                    // Default rate if no service+env pair matches
+                    sample_rate_zv = zend_hash_str_find(DDTRACE_G(agent_rate_by_service), ZEND_STRL("service:,env:"));
```

Comment:
Seems a bit odd to look up `"service:,env:" in a hash this way, but I'm not against it if that's considered easier or simpler for something.